Allow dumping seccomp filters of existing processes

mephi42 opened the pull request:
This patch adds -p option to dump subcommand (like the one gdb, strace,
etc support) that uses PTRACE_SECCOMP_GET_FILTER (available since Linux
4.4), which allows dumping seccomp filters of existing processes.

david942j commented:
Thanks for the contribution! This is a great feature to have.
I haven't tried the changes but need to point out something first:
1. Make sure you pass the linter (rubocop) and the tests
  * `bundle exec rake`
2. Add tests against this new future and ensure 100% coverage
  * the percentage of coverage should be shown after you run the tests

david942j commented on lib/seccomp-tools/dumper-pid.rb:
```diff
@@ -0,0 +1,42 @@
+    #   #=> []
+    #   dump(pid2, 1) { |c| c[0, 10] }
+    #   #=> [" \x00\x00\x00\x00\x00\x00\x00\x15\x00"]
+    def dump(pid, limit, &block)
```
I would prefer to add a function (`dump_by_pid`) in `module Dumper` instead of create a new module.

david942j commented on lib/seccomp-tools/dumper-pid.rb:
```diff
@@ -0,0 +1,42 @@
+    #   dump(pid2, 1) { |c| c[0, 10] }
+    #   #=> [" \x00\x00\x00\x00\x00\x00\x00\x15\x00"]
+    def dump(pid, limit, &block)
+      Ptrace.attach(pid)
```
Please check whether the ptrace attach succeeds. It's normal that user may pass a pid that they don't have permission to attach on. Detect this scenario and print the possible solution to the users.

mephi42 commented:
Thanks for the review! I've moved the dumper function to the existing module, and added the error handling.

With tests it's more complicated, since the new function requires `CAP_SYS_ADMIN`. Would it make sense to have a test that is skipped when current user lacks this capability?

david942j commented on lib/seccomp-tools/dumper.rb:
```diff
@@ -129,5 +129,39 @@ def handle_child(*args)
+    #   Return the block returned. If block is not given, array of raw bytes will be returned.
+    # @example
+    #   dump(pid1, 1)
+    #   #=> []
```
add comment about why this is supposed to be empty (`pid1` is a process without seccomp installed or?)
Also suggest to do launch a process (through popen or Open3) and use its pid as examples here. Binaries under `spec/binary/` can be used.

mephi42 commented:
I now have 100% coverage when running the test suite as root. I've also added a small fix for a false positive, which was getting in the way, as well as clean ptrace detachment. I didn't try tweaking Travis config yet.

david942j commented on lib/seccomp-tools/cli/dump.rb:
```diff
@@ -48,6 +49,10 @@ def parser
                    option[:ofile] = o
                  end
+
+          opt.on('-p', '--pid PID', 'Dump seccomp filters of the existing process.', Integer) do |p|
```
Mention that the user has to be root to use this option

david942j commented on lib/seccomp-tools/cli/dump.rb:
```diff
@@ -56,14 +61,26 @@ def parser
+          rescue Errno::EPERM, Errno::EACCES => e
+            warn(e)
+            warn('PTRACE_SECCOMP_GET_FILTER requires CAP_SYS_ADMIN')
+            warn("Try this: sudo seccomp-tools dump -p #{option[:pid]} -l #{option[:limit]}")
```
replace the option here to be exactly what the user passed to prevent confusing the user

david942j commented on spec/dumper_spec.rb:
```diff
@@ -58,7 +58,7 @@
         expect(got.size).to be bpf.size
         expect(got[0, 0x1ec]).to eq bpf[0, 0x1ec]
-        expect(got[0x1ee..-1]).to eq bpf[0x1ee..-1]
+        expect(got[0x1f0..-1]).to eq bpf[0x1f0..-1]
```
👍

mephi42 commented:
The test failures are most likely due to temporary privilege drop - some rubies must be reading library files in advance, and some not. I will change the test to do the least amount of work possible when running with reduced privileges.

david942j commented:
Good job, thanks!
